third_party/move/move-vm/types/src/interner.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Aptos Foundation
```

Comment:
So this is a generic and efficient interner implementation, inspired by https://matklad.github.io/2020/03/22/fast-simple-rust-interner.html

It does has unsafe code, but it's the same technique also used by [typed_arena](https://crates.io/crates/typed_arena), which is already in our VM. 

Let me know if you'd prefer me moving this into a separate crate for better isolation though.